### ISSUE TITLE
[1.19.2 Fabric] Drowned Guardian Jockey mixin will deadlock world generation.

### ISSUE THREAD

**Issue by TelepathicGrunt** [sf-inc/sihywtcamd#38]
I am the dev for Repurposed Structures and looked into a deadlock reported to me with All of Fabric 6 modpack. The user was finding the game was locking up and staying stuck forever when certain chunks are attempted to be generated.

I went ahead and performed a thread dump and found this: https://controlc.com/1809dda0

Basically, this code is unsafe and should never be ran during world generation. The reason why is because the chunk is currently generating, tries to spawn a structure, structure tries to spawn a Drowned as part of it, this mod's mixin runs and tries to add Drowned as passenger to Guardian, triggers the onDismounted method, which then calls the getBlockState method for the entity's ServerLevel instead of the WorldGenRegion, now code path is locked up waiting for the currently generating chunk to finish generating except that is itself so it is stuck waiting on itself.
https://github.com/sf-inc/sihywtcamd/blob/c2648495d8c29ca7a7eb47bf26ef359538936846/src/main/java/com/github/galatynf/sihywtcamd/mixin/zombie/DrownedMixin.java#L52

Perhaps you can replace this code with directly setting the passenger and vehicle fields instead in a way that is safe for worldgen. Or look into how vanilla is able to spawn entities with riders safely from nbt files during worldgen. You can test your fix by running your mod with my Repurposed Structures mod, use /locate command to find my Jungle Fortress that spawns Drowned when tunnels are below sea level, and teleport to it. If all the chunks around generates properly and nothing is stuck, the fix works

**Comment by TelepathicGrunt** [comment]
Looks like all your jockey code are running the same method and all are unsafe for worldgen jockey spawning: 
https://github.com/sf-inc/sihywtcamd/blob/c2648495d8c29ca7a7eb47bf26ef359538936846/src/main/java/com/github/galatynf/sihywtcamd/mixin/spider/CaveSpiderMixin.java#L33

**Comment by charlyb01** [comment]
Hello, thanks for your report! I guess it's the same issue I had with Unstructured on #36. At that time, I didn't understand what would be the issue as I was doing the same thing I found on Minecraft code but it seems I missed something. I will check again soon, thanks again!

**Comment by noncom** [comment]
Hey so I've got the same bug when building myself a modpack.
In my case it seems like it might be a conflict between Lithium, Phosphor and Sihywtcamd.
But probably Lithium and Phosphor might be not strictly necessary, idk.
I will update this comment if I have more info on that, a bit later (if/when time allows).
But you could try to reproducing this yourself!

Steps for reproduction:
- Load my modpack (attached in the end of this message)
- Create a world with seed "test", cheats: on, difficulty: hard. Leave all other settings as is, including the BetterX world type
- Start the game, you should spawn in a green area
- Upon spawning set gamemode to creative
- Enter the following lines to the chat, copying one by one and pressing enter after each:
```lua
#allowdiagonalascend true
#allowdiagonaldescend true
#allowparkour true
#goal 10000 10000 10000
#path
```
  I think you know what is Baritone, but just in case: you should see that the bot confirms all these commands. You don't need the `/` at the start! Just send these messages into the chat.
- Observe how Baritone navigates you through the world for a while.
- Upon reaching the point shown on the screenshot, roughly about 3400, 72, 3400 you shall encounter this bug and find yourself at the spot shown in the screenshot


![2023-06-15_20 05 58](https://github.com/sf-inc/sihywtcamd/assets/2920144/ab7b3da1-34fb-48b9-a718-acf6a88108e6)

Here I'm also attaching a snapshot from jvisualvm. You can load it in jvisualvm and see the situation: 
[snapshot-1686852055004.zip](https://github.com/sf-inc/sihywtcamd/files/11761305/snapshot-1686852055004.zip)

Screenshots of the snapshot that give some idea of the important points:
1. 
![image](https://github.com/sf-inc/sihywtcamd/assets/2920144/130563b6-36fe-40c4-a8c3-fb0d1fb749e4)
2. 
<img width="833" alt="image" src="https://github.com/sf-inc/sihywtcamd/assets/2920144/6ef5d78b-2801-4b0e-92b7-263bdb414969">


Modpack:
[Journey's End Light-1.4.zip](https://github.com/sf-inc/sihywtcamd/files/11761286/Journey.s.End.Light-1.4.zip)

**Comment by TelepathicGrunt** [comment]
This bug has nothing to do with any other mod. The bug is specifically a mixin into drowned spawning code and calling a method that is not worldgen safe. Drowned saved as part of a structure nbt file will spawn during worldgen and trigger this mixin to run. That’s what the deadlock is. No need for lithium or whatever. They play no role in this

**Comment by noncom** [comment]
Oh great! Than that's the explanation.
Let's hope it's fixed asap then! 😄

**Comment by arazadaz** [comment]
I'm experiencing this as well in my modpack. Traced it back to this mod. Telepathic Grunt is entirely correct.

**Comment by noncom** [comment]
Yeah. Just a notice: I've left the modpack and reproduction info because it's a 100% reproduction case.
Might be used for verifying the solution if needed.

**Comment by charlyb01** [comment]
Thank you for your additional information. I already knew this bug thanks to the issue though, so I will be working on fixing this as soon as I have time. The fix should come with next version =)

**Comment by charlyb01** [comment]
Thank you all for your help, this was really helpful =)
This is fixed on 1.20.1 (and newer) versions

**Comment by izofar** [comment]
@charlyb01 I've found this issue due to a crash report on my mod page for Bygone Nether. When you mention fixing it for 1.20.1, do you not support previous versions? If not, I will list as an incompatibility with mine for those versions.
